wikidatarefisland/scraper.py

```diff
@@ -0,0 +1,84 @@
+    def __init__(self, config, schemaorg_normalizer, schemaorg_mapper):
+        """
+
+        :type config: wikidatarefisland.Config
+        :type schemaorg_mapper: wikidatarefisland.services.SchemaorgPropertyMapper
+        """
+        self.config = config
+        self.schemaorg_normalizer = schemaorg_normalizer
+        self.schemaorg_mapper = schemaorg_mapper
+
+    def flow(self, item):
+        resource_urls = {i['url']: i['referenceMetadata'] for i in item['resourceUrls']}
+        schemaorg_mapping = {
+            i['url']: i['property'] for i in self.schemaorg_mapper.get_mapping()}
+        statement_pids = [i['pid'] for i in item['statements']]
+        properties_to_check = filter(lambda t: t in schemaorg_mapping.values(), statement_pids)
+        if not properties_to_check:
+            # Bail out if we can't find any property that have schema.org equivalent
+            return
+
+        extracted_data = {}
+        with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
+            future_to_url = {
+                executor.submit(self._check_external_identifier, i): i for i in resource_urls}
+        for future in concurrent.futures.as_completed(future_to_url):
+            url = future_to_url[future]
+            try:
+                data = future.result()
+            except KeyboardInterrupt:
+                raise
+            except Exception as exc:
+                print('%r generated an exception: %s' % (url, exc))
+            else:
+                extracted_data[url] = data
+
+        return self._match_extracted_data(extracted_data, resource_urls,
+                                          schemaorg_mapping, item)
+
+    def _check_external_identifier(self, url):
+        r = requests.get(url, timeout=30, user_agent=self.config.get('user_agent'))
+        base_url = get_base_url(r.text, r.url)
+        data = extruct.extract(r.text, base_url=base_url, uniform=True)
+        return {'data': self.schemaorg_normalizer.normalize_from_extruct(data),
+                'timestamp': strftime("%Y-%m-%d %H:%M:%S", gmtime())}
+
+    def _match_extracted_data(self, extracted_data, resourceUrls, mapping, item):
+        final_data = []
+        for url in extracted_data:
+            resourceUrls[url]['dateRetrieved'] = extracted_data[url]['timestamp']
+            for schema_property in extracted_data[url]['data']['properties']:
+                if schema_property not in mapping:
+                    continue
+                pid = mapping[schema_property]
+                for statement in item['statements']:
+                    if statement['pid'] != pid:
+                        continue
+
+                    formatted = self._format_result(
+                        statement,
+                        item['itemId'],
+                        resourceUrls[url],
+                        extracted_data[url]['data']['properties'][schema_property])
+                    final_data.append(formatted)
```

Comment:
Perhaps it's a matter of personal style but IMHO the nesting runs a little too deep in this part, which can cause a bit of confusion. There are a couple of ways to mitigate that:

1. Splitting this into three different functions, each of which represents a single iterator, this take care of the nesting
1. Using reducers instead of `for in`s  can take care of the mutability of `final_data` so that we don't have to depend on it (and pass it down to each function by reference)

Reply:
yeah, I need to split this to smaller methods but I'm not sure it would be okay to split it to several pipes. Let me take a stab at it.